### PR TITLE
Simplify Endpoint Deployment and Management in API Gateway

### PR DESCRIPTION
**Pull request by uklok-agent**
## Description

To simplify the process of deploying new endpoints into our API gateway, we've built a straightforward toolkit. This set of tools helps create, fund, and update subscriptions efficiently.

### Key Features

- **Chainlink Functions Toolkit**: Streamlines the creation, funding, and updating of subscriptions connected to the gateway.
- **Endpoint Management Tools**: Allows easy creation and updating of new endpoints.
- **Hardhat Tasks**: Automates the process of upserting these endpoints into the gateway.
- **NPM Scripts**: Provides easy access to these tasks for everyone.

### How to Add a New Endpoint

1. **Test Your Function**:
   - Use the [Chainlink Functions Playground](https://functions.chain.link/playground) to ensure your function works as expected.

2. **Create a New File**:
   - Follow the naming pattern `<next_number>_<function_name>.[js|ts]` in the `packages/hardhat/functions` folder.
   - Paste the working code from the playground into this new file.

3. **Deploy the Endpoint**:
   - Run the command: `npm run link:gw:upsert-fund --network <network_name>` in the Hardhat project.
   - This deploys the new endpoint to the gateway on the specified network.

### How to Use the Endpoint

1. **Find Endpoint Details**:
   - Check the `link_functions_map.json` file next to your function for the `subscriptionId` and `consumerAddress`.

2. **Integrate in Your Smart Contract**:
   - Use the `consumerAddress` to create a gateway instance with the `ICCGateway` interface.
   - Use the `sendRequest` method with the `subscriptionId` to access the endpoint.

### Updating an Endpoint

1. **Edit the Function**:
   - Modify the code in the existing function file.

2. **Update the Gateway**:
   - Run: `npm run link:gw:upsert`
   - This updates the endpoint in the gateway, making your changes live immediately.

### Security Note

For security reasons, the gateway uses an ACL. To publish, update endpoints, or include new clients, your EOA must have a MANAGER role.

---

### Additional Information

- The funding process aims to distribute LINK tokens evenly in the `deployer` wallet, ensuring each subscription (endpoint) receives between 3 and 10 LINK tokens.
- The `@chainlink/functions-toolkit` is currently using ethers@v5, which has several incompatibilities with the project's ethers@v6. Various workarounds were implemented to address these issues.
- New interfaces extending Hardhat were created to include Chainlink network details.
- The list of potential networks for deployment was updated to include Chainlink-specific details.
- The Gateway and an ExampleClient smart contract have already been deployed on `Polygon Mumbai` and `Optimism Sepolia`.

---
Your ENS/address: [uklok.eth](https://app.ens.domains/uklok.eth)

### CODE REVIEW

File: packages/hardhat/functions/00_backstory_gpt.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+const CHARACTER_NAME = args[2]; // --------------> |   Eldon
+// [2] PROMPT ENGINEERING //
+
+const prompt = `Generate a D&D 5e backstory for ${CHARACTER_NAME}, a Level 1 ${CHARACTER_CLASS} who is a ${CHARACTER_RACE}. The backstory should be extremely short, dense, powerful, engaging, coherent, and suitable for a fantasy RPG setting. Just reply with the one paragraph backstory and nothing else.`;
+
+// [3] AI DATA REQUEST //
+
+// requests: OpenAI API using Functions
+const openAIRequest = await Functions.makeHttpRequest({
```
Comment (mayanloong) [inline packages/hardhat/functions/00_backstory_gpt.ts:17]: Is this await required? It seems that await the request twice?

**Review by mayanloong — APPROVED** [review]
confirm

File: packages/hardhat/functions/00_backstory_gpt.ts (hunk 1 above)
Comment (tippi-fifestarr) [reply packages/hardhat/functions/00_backstory_gpt.ts:17]: let me check

Comment (tippi-fifestarr) [reply packages/hardhat/functions/00_backstory_gpt.ts:17]: yes that is redundant. can probably just assign response = await Functions etc, ty

**Comment by tippi-fifestarr** [comment]
I ran it and looked over what you did, from what I understand it looks like you saved a bunch of future developers a bunch of redundant code. How can I check it works, there are provided examples but not sure how to use them and adapt to my [current character generator contract](https://github.com/tippi-fifestarr/Ceptor-scaffold-OP/blob/character-generation-contracts/packages/hardhat/contracts/CharacterGen.sol), i imagine I can add a function or import an interface... also, i'm confused about endpoints, how does this relate to what i should be doing if i were not using gateway.

As a Chainlink Dev Expert, I'm going to have to learn to do it the normal way before I can really appreciate any savings and how to use this gateway. This is super cool, makes me think about if the gateway can work for other reusable components

**Review by tippi-fifestarr — APPROVED** [review]
awesome! above my current technical level, i'll delay my functions integration until last minute so i can use this if possible teamwork makes the dreamwork 👯‍♂️

**Pull request merged**